gossipsub_test.go

```diff
@@ -23,6 +23,26 @@ func getGossipsubs(ctx context.Context, hs []host.Host, opts ...Option) []*PubSu
 	return psubs
 }
 
+func waitForMeshConstruction(topic string, psubs []*PubSub) {
+	for _, ps := range psubs {
+		grt := ps.rt.(*GossipSubRouter)
+		for len(grt.mesh[topic]) < GossipSubDlo {
+			time.Sleep(100 * time.Millisecond)
+		}
+	}
+}
+
+// wait for all incoming subscription messages to be processed
+// this method should ONLY be called if all connected peers subscribe to the same topic
+func waitForSubscriptionProcessing(topic string, psubs []*PubSub) {
```

Comment:
Races with PubSub.handleIncomingRPC when it tries to update Pubsub.topics[T].,

Reply:
See above: PubSub is currently setup to basically have a few long running goroutines with event loops in them instead of using mutexes. There's an internal `p.eval` channel that will run arbitrary functions for you inside of the event loop so you don't run into races.

Reply:
Aha... that is brilliant. Let me fix this.